hivemq-edge/src/main/java/com/hivemq/protocols/ProtocolAdapterManager.java

```diff
@@ -127,92 +131,105 @@ public ProtocolAdapterManager(
         this.jsonPayloadDefaultCreator = jsonPayloadDefaultCreator;
         this.protocolAdapterWritingService = protocolAdapterWritingService;
         this.executorService = executorService;
-        this.protocolAdapterTagService = protocolAdapterTagService;
+        protocolAdapterFactoryManager = new ProtocolAdapterFactoryManager(moduleLoader, eventService, protocolAdapterWritingService.writingEnabled());
     }
 
 
     @SuppressWarnings({"unchecked", "rawtypes"})
     public @NotNull ListenableFuture<Void> start() {
 
-        protocolAdapterWritingService.addWritingChangedCallback(this::findAllAdapters);
-        findAllAdapters();
-
-        //iterate configs and start each adapter
-        final Map<String, Object> allConfigs =
-                configurationService.protocolAdapterConfigurationService().getAllConfigs();
+        log.info("Starting adapters");
+        protocolAdapterWritingService.addWritingChangedCallback(() -> {
+            protocolAdapterFactoryManager.writingEnabledChanged(protocolAdapterWritingService.writingEnabled());
+        });
 
-        if (allConfigs.isEmpty()) {
-            return Futures.immediateFuture(null);
-        }
         final ImmutableList.Builder<CompletableFuture<Void>> adapterFutures = ImmutableList.builder();
 
-        for (final Map.Entry<String, Object> configSection : allConfigs.entrySet()) {
+        for (final Map.Entry<String, Object> configSection : configPersistence.allAdapters().entrySet()) {
             final String adapterType = getKey(configSection.getKey());
-            final ProtocolAdapterFactory<?> protocolAdapterFactory = getProtocolAdapterFactory(adapterType);
-            if (protocolAdapterFactory == null) {
-                if (log.isWarnEnabled()) {
-                    log.warn("Protocol adapter for config {} not found.", adapterType);
-                }
-                continue;
+
+            final Optional<ProtocolAdapterFactory<?>> protocolAdapterFactoryOptional = protocolAdapterFactoryManager
+                    .get(adapterType); //FIXME should be used as actual optional
```

Comment:
FIXME

Reply:
I will remove the comment.

The class has tons of places where we either don't use Optionals correctly or micing sync with CompletableFuture-based operations.

That will be something for the next iteration.